omegaconf/base.py

```diff
@@ -626,8 +666,10 @@ def resolve_parse_tree(
         """
         from .nodes import StringNode
 
-        def node_interpolation_callback(inter_key: str) -> Optional["Node"]:
-            return self._resolve_node_interpolation(inter_key=inter_key)
+        def node_interpolation_callback(
+            inter_key: str, memo: Set[int]
```

Comment:
Probably that `quoted_string_callback` and `resolver_interpolation_callback` should also use `memo` (for the latter, most resolvers won't use it, but we could at least support it in relevant builtin resolvers like `oc.decode`)

Reply:
Just to expand a bit on resolvers, what I have in mind is adding a new `_memo_` resolver argument (using the same mechanism as `_parent_` and `_root_`). It could be used by `oc.decode`, `oc.dict.{keys,values}` and `oc.select` for instance.

Reply:
Your new quoted strings tests aren't actually testing quoted strings, they're all top level strings. You can use `oc.decode` to test quoted strings, e.g.:
```python
cfg = OmegaConf.create({"a": "${oc.decode:'x ${a}'}"})
cfg.a
```

On resolvers: I think it could remain documented and have a docstring that says "internal use only" (or we could create a separate private register function). But I'm fine with keeping it for later.

Reply:
Quoted strings: Added a proper test and fixed.

Custom resolvers: 
Feel free to introduce it when we have a custom resolver that can actually benefit from it.
Standard usage of custom resolvers does not have this issue because they are not the ones doing derefencing the interpolations.

Reply:
> Custom resolvers:
> Feel free to introduce it when we have a custom resolver that can actually benefit from it.
> Standard usage of custom resolvers does not have this issue because they are not the ones doing derefencing the interpolations.

To be honest I probably won't bother, but right now `oc.decode` could already benefit from it since it's able to resolve interpolations.

Reply:
Can you show oc.decode usage that is failing to detect an interpolation cycle with this PR?

Reply:
```python
cfg = OmegaConf.create({"a": "${oc.decode:${oc.env:VAR}}", "b": r"${oc.decode:'\${b}'}"})
os.environ["VAR"] = "${a}"
cfg.a  # or cfg.b
```

The case with `cfg.a` shows how it could happen in practice. The case with `cfg.b` is what I would put in a unit test.